File: src/maya.lib/haab.fs
namespace Maya

(*0-19*)
type HaabDay = int

[<Struct>]
type HaabMonth =
    | Pop
    | Wo
    | Sip
    | Sotz
    | Sek
    | Xul
    | Yaxkin
    | Mol
    | Chen
    | Yax
    | Sak
    | Keh
    | Mak
    | Kankin
    | Muwan
    | Pax
    | Kayab
    | Kumku
    | Wayeb

type HaabDate = (struct (HaabDay * HaabMonth))

module Haab =

    let ofNumber n = n |> function
        | 01 -> Pop
        | 02 -> Wo
        | 03 -> Sip
        | 04 -> Sotz
        | 05 -> Sek
        | 06 -> Xul
        | 07 -> Yaxkin
        | 08 -> Mol
        | 09 -> Chen
        | 10 -> Yax
        | 11 -> Sak
        | 12 -> Keh
        | 13 -> Mak
        | 14 -> Kankin
        | 15 -> Muwan
        | 16 -> Pax
        | 17 -> Kayab
        | 18 -> Kumku
        | 19 -> Wayeb
        | _ -> invalidArg "n" "n must be between 1-19"
        
    let number month = month |>  function
        | Pop -> 1
        | Wo -> 2
        | Sip -> 3
        | Sotz -> 4
        | Sek -> 5
        | Xul -> 6
        | Yaxkin -> 7
        | Mol -> 8
        | Chen -> 9
        | Yax -> 10
        | Sak -> 11
        | Keh -> 12
        | Mak -> 13
        | Kankin -> 14
        | Muwan -> 15
        | Pax -> 16
        | Kayab -> 17
        | Kumku -> 18
        | Wayeb -> 19

    let meaning month = month |> function
        | Pop -> "mat"
        | Wo -> "black conjunction"
        | Sip -> "red conjunction"
        | Sotz -> "bat"
        | Sek -> "death"
        | Xul -> "dog"
        | Yaxkin -> "new sun"
        | Mol -> "water"
        | Chen -> "black storm"
        | Yax -> "green storm"
        | Sak -> "white storm"
        | Keh -> "red storm"
        | Mak -> "enclosed"
        | Kankin -> "yellow sun"
        | Muwan -> "owl"
        | Pax -> "planting time"
        | Kayab -> "turtle"
        | Kumku -> "granary"
        | Wayeb -> ""

    let construct (date: LongDate): HaabDate =

        let days = LongDate.daysSinceEpoch date
    
        (*-3113 BCE September 9th, 13.0.0.0.0 4 Ajaw, 8 Kumk’u *)
        (* 2012 12 21 4 Ajaw, 3 K'ank'in*)

        let head =
            match LongDate.Epoch with
            | BC3114 ->    
                (8 + ((number Kumku - 1) * 20))
            | CE2012 ->
                ((3 + ((number Kankin - 1) * 20)))

        let dayOfHaab = (days + head) % 365

        let day =
            dayOfHaab % 20
            |> int
        let month = 
            (dayOfHaab/20) + 1
            |> ofNumber

        struct (day, month)
 
    let ofDate y m d =
        let date = LongDate.construct y m d
        construct date

    let ofDateOnly (date: System.DateOnly): HaabDate =
        let (y, m, d) = date.Deconstruct()
        ofDate y m d

    

type HaabMonth with
        member month.Meaning = Haab.meaning month
        member month.Number = Haab.number month

File: code/rust/maya/src/haab.rs
use core::panic;

use crate::long::LongDate;

pub type HaabDay = i64;

#[derive(Debug)]
pub enum HaabMonth {
    Pop,
    Wo,
    Sip,
    Sotz,
    Sek,
    Xul,
    Yaxkin,
    Mol,
    Chen,
    Yax,
    Sak,
    Keh,
    Mak,
    Kankin,
    Muwan,
    Pax,
    Kayab,
    Kumku,
    Wayeb
}

pub type HaabDate = (HaabDay, HaabMonth);

impl HaabMonth {
    pub fn of_number(n: i64) -> HaabMonth {
        match n {
            00 => HaabMonth::Wayeb,
            01 => HaabMonth::Pop,
            02 => HaabMonth::Wo,
            03 => HaabMonth::Sip,
            04 => HaabMonth::Sotz,
            05 => HaabMonth::Sek,
            06 => HaabMonth::Xul,
            07 => HaabMonth::Yaxkin,
            08 => HaabMonth::Mol,
            09 => HaabMonth::Chen,
            10 => HaabMonth::Yax,
            11 => HaabMonth::Sak,
            12 => HaabMonth::Keh,
            13 => HaabMonth::Mak,
            14 => HaabMonth::Kankin,
            15 => HaabMonth::Muwan,
            16 => HaabMonth::Pax,
            17 => HaabMonth::Kayab,
            18 => HaabMonth::Kumku,
            19 => HaabMonth::Wayeb,
            _ => panic!()
        }
    }
    
    pub fn number(&self) -> i64 {
        match self {
            HaabMonth::Pop => 1,
            HaabMonth::Wo => 2,
            HaabMonth::Sip => 3,
            HaabMonth::Sotz => 4,
            HaabMonth::Sek => 5,
            HaabMonth::Xul => 6,
            HaabMonth::Yaxkin => 7,
            HaabMonth::Mol => 8,
            HaabMonth::Chen => 9,
            HaabMonth::Yax => 10,
            HaabMonth::Sak => 11,
            HaabMonth::Keh => 12,
            HaabMonth::Mak => 13,
            HaabMonth::Kankin => 14,
            HaabMonth::Muwan => 15,
            HaabMonth::Pax => 16,
            HaabMonth::Kayab => 17,
            HaabMonth::Kumku => 18,
            HaabMonth::Wayeb => 19,
        }
    }

    pub fn meaning(&self) -> &'static str {
        match self {
            HaabMonth::Pop => "mat",
            HaabMonth::Wo => "black conjunction",
            HaabMonth::Sip => "red conjunction",
            HaabMonth::Sotz => "bat",
            HaabMonth::Sek => "death",
            HaabMonth::Xul => "dog",
            HaabMonth::Yaxkin => "new sun",
            HaabMonth::Mol => "water",
            HaabMonth::Chen => "black storm",
            HaabMonth::Yax => "green storm",
            HaabMonth::Sak => "white storm",
            HaabMonth::Keh => "red storm",
            HaabMonth::Mak => "enclosed",
            HaabMonth::Kankin => "yellow sun",
            HaabMonth::Muwan => "owl",
            HaabMonth::Pax => "planting time",
            HaabMonth::Kayab => "turtle",
            HaabMonth::Kumku => "granary",
            HaabMonth::Wayeb => ""
        }
    }
}

pub fn construct(date: &LongDate) -> HaabDate {
    let days = date.days_since_epoch;
    let head = match date.epoch {
        crate::MayaEpoch::BC3114 =>
            8 + ((HaabMonth::Kumku.number()  - 1) * 20),
        crate::MayaEpoch::CE2012 =>
            3 + ((HaabMonth::Kankin.number() - 1) * 20)
    };
    let dayOfHaab = (days + head) % 365;
    let day = dayOfHaab % 20;
    let month = (dayOfHaab/20) + 1;
    let haabMonth = HaabMonth::of_number(month);
    (day, haabMonth)
}